src/rust/src/http_client.rs

```diff
@@ -48,26 +59,32 @@ impl BufferedHttpClient {
             head: 0,
         }
     }
+
     pub async fn get(&mut self, begin: usize, length: usize, min_req_size: usize) -> Result<&[u8]> {
-        let tail = self.head + self.buf.len();
-        if begin + length > tail || begin < self.head {
+        // If we don't already have the data, we download it
+        if begin + length > self.tail() || begin < self.head {
             // Remove bytes before new begin
-            if begin > self.head && begin < tail {
+            if begin > self.head && begin < self.tail() {
                 let _ = self.buf.split_to(begin - self.head);
                 self.head = begin;
-            } else if begin >= tail || begin < self.head {
+            } else if begin >= self.tail() || begin < self.head {
                 self.buf.clear();
                 self.head = begin;
             }
 
             // Read additional bytes
-            let range_begin = max(begin, tail);
-            let range_length = max(begin + length - range_begin, min_req_size);
+            let range_begin = max(begin, self.tail());
+            let range_length = max(length - (range_begin - begin), min_req_size);
```

Comment:
This change here is critical. Don't understand it immediately. Will add some comments to my code, to understand it.

Reply:
While tracking down the `stale tail` bug, I re-wrote this line in a way that made more sense to me. 

It's mathematically equivalent, but unnecessary. I'll revert it momentarily.

Reply:
Just to be clear, the bug fixed was the line before that: 

```
            let range_begin = max(begin, tail);
```

Is invalid if we've reset head (e.g. https://github.com/bjornharrtell/flatgeobuf/pull/93/files/003a76442ec9366491eaa1ed6fc1097890ca9c42#diff-75c21172adb6c41e0eb130a61d0bdd00983afafab292cfbc811844d964f21abeR69), then `tail` is no longer valid.